execution/executor/src/components/block_tree/mod.rs

```diff
@@ -0,0 +1,257 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+#![forbid(unsafe_code)]
+
+#[cfg(test)]
+mod test;
+
+use crate::logging::{LogEntry, LogSchema};
+use anyhow::{anyhow, ensure, Result};
+use consensus_types::block::Block as ConsensusBlock;
+use diem_crypto::HashValue;
+use diem_infallible::Mutex;
+use diem_logger::{debug, info};
+use diem_types::{ledger_info::LedgerInfo, proof::definition::LeafCount, protocol_spec::DpnProto};
+use executor_types::{Error, ExecutedChunk};
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    sync::{Arc, Weak},
+};
+use storage_interface::DbReader;
+
+pub struct Block {
+    pub id: HashValue,
+    pub output: ExecutedChunk,
+    children: Mutex<Vec<Arc<Block>>>,
+    block_lookup: Arc<BlockLookup>,
+}
+
+impl Drop for Block {
+    fn drop(&mut self) {
+        self.block_lookup.remove(self.id);
+        debug!(
+            LogSchema::new(LogEntry::SpeculationCache).block_id(self.id),
+            "Block dropped."
+        );
+    }
+}
+
+impl Block {
+    fn add_child(&self, child: Arc<Self>) {
+        self.children.lock().push(child)
+    }
+
+    pub fn num_persisted_transactions(&self) -> LeafCount {
+        self.output.result_view.txn_accumulator().num_leaves()
+    }
+
+    fn ensure_has_child(&self, child_id: HashValue) -> Result<()> {
+        ensure!(
+            self.children.lock().iter().any(|c| c.id == child_id),
+            "{:x} doesn't have child {:x}",
+            self.id,
+            child_id,
+        );
+        Ok(())
+    }
+}
+
+fn epoch_genesis_block_id(ledger_info: &LedgerInfo) -> HashValue {
+    ConsensusBlock::make_genesis_block_from_ledger_info(ledger_info).id()
+}
+
+struct Inner(HashMap<HashValue, Weak<Block>>);
+
+impl Inner {
+    fn multi_get(&self, ids: &[HashValue]) -> Result<Vec<Option<Arc<Block>>>> {
+        ids.iter()
+            .map(|id| {
+                self.0
+                    .get(id)
+                    .map(|weak| {
+                        weak.upgrade()
+                            .ok_or_else(|| anyhow!("Block {:x} has been deallocated.", id))
+                    })
+                    .transpose()
```

Comment:
this is bizarre

Reply:
which part? suggestion?

Reply:
the whole function, I'd just use less functional style for more readability